### PR TITLE
gui: improve keyboard control

### PR DESCRIPTION
**Pull request by pieleric**

### CODE REVIEW

**Review by Sabrina85 — CHANGES_REQUESTED** [review]

File: src/odemis/gui/comp/canvas.py [hunk 1]
```diff
@@ -1608,8 +1608,9 @@ def on_char(self, evt):
                 return
 
             self.shift_view(move_px)
-        else:
-            super(DraggableCanvas, self).on_char(evt)
+            return
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/gui/comp/canvas.py:1611]: It's not per se something for the docstring. All event handlers follow more or less the same behaviour. However, here, when we find that the user pressed a key which we are interested in (and process) we have to indicate this key is officially handled, by immediately returning (and not letting other code process it further). I've added a comment.

File: src/odemis/gui/comp/viewport.py [hunk 2]
```diff
@@ -1356,80 +1352,113 @@ def on_vrange(self, vrange):
...
+        wx.WXK_UP: 0.1,
+        wx.WXK_DOWN:-0.1,
+    }
+    _key_to_hscale = {
+        ord("*"):-0.1,
+        ord("/"):0.1,
+    }
+    _key_to_vscale = {
+        ord("-"):0.1,
+        ord("+"):-0.1,
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/gui/comp/viewport.py:1370]: Yes, you understood correctly. I've add some extra comments.

File: src/odemis/gui/comp/viewport.py [hunk 3]
```diff
@@ -1356,80 +1352,113 @@ def on_vrange(self, vrange):
...
+    }
+
+    def on_char(self, evt):
+        key = evt.GetKeyCode()
+
+        if key in self._key_to_hmove:
+            move = self._key_to_hmove[key]
+            if evt.ShiftDown():  # softer
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/gui/comp/viewport.py:1378]: No, we don't support increasing the stepsize. For the value inputs, we do support it (with Ctrl), but typically Ctrl+KEY is for shortcuts, so it's a little tricky to use Ctrl in every case. So for now we don't support it. No one ever asked for it either...

File: src/odemis/gui/comp/viewport.py [hunk 4]
```diff
@@ -1356,80 +1352,113 @@ def on_vrange(self, vrange):
...
+                move *= 0.1
+
+            # convert ratio to units
+            rng = self.hrange.value
+            if rng is None:
+                return
+            shift = (rng[1] - rng[0]) * move
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/gui/comp/viewport.py:1385]: I've added a docstring which should summarize this.

File: src/odemis/gui/comp/viewport.py [hunk 5]
```diff
@@ -1356,80 +1352,113 @@ def on_vrange(self, vrange):
...
+            if rng is None:
+                return
+
+            centre = rng[0]  # Keep the bottom at the bottom
+            self.rescale_vertical(scale, centre)
+        else:
+            evt.Skip()
+
+    def on_hlegend_scroll(self, evt):
+        """ Scroll event for the bottom legend.
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/gui/comp/viewport.py:1428]: Yes. That's the .bottom_legend attribute of the object.

File: src/odemis/gui/comp/viewport.py [hunk 6]
```diff
@@ -1356,80 +1352,113 @@ def on_vrange(self, vrange):
...
         """
-        if not self.canvas.has_data():
-            return
 
         rot = evt.GetWheelRotation() / evt.GetWheelDelta()
-        rng = self.vrange.value
-        span = rng[1] - rng[0]
-        zoom_centre = self.canvas.pos_y_to_val_y(evt.Position[1])
-        scale = 0.1
+        scale = -0.1 * rot
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/gui/comp/viewport.py:1434]: 0.1 means 10% of the range, like for the key press. I've added a comment.

File: src/odemis/gui/comp/viewport.py [hunk 7]
```diff
@@ -1356,80 +1352,113 @@ def on_vrange(self, vrange):
...
+        if evt.ShiftDown():  # softer
+            scale *= 0.1
 
-        # proportion of zoom_centre
-        prop = (zoom_centre - rng[0]) / span
+        zoom_centre = self.canvas.pos_x_to_val_x(evt.Position[0])
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/gui/comp/viewport.py:1438]: Well, the goal is to zoom with the position under the mouse staying fixed. So we set the current mouse position as the "zoom centre".

File: src/odemis/gui/comp/viewport.py [hunk 8]
```diff
@@ -1356,80 +1352,113 @@ def on_vrange(self, vrange):
...
+        self.rescale_horizontal(scale, zoom_centre)
 
-        # zoom around the centre point
-        if rot < 0:
-            new_span = span * (1 + scale)
-        else:
-            new_span = span * (1 - scale)
+    def on_vlegend_scroll(self, evt):
+        """ Scroll event for the left legend.
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/gui/comp/viewport.py:1442]: The left legend is the y-axis. It's already indicate in the line just below.

File: src/odemis/gui/comp/viewport.py [hunk 9]
```diff
@@ -1554,6 +1549,121 @@ def on_canvas_motion(self, evt):
...
+        # Clamp the ranges - this functions differently for scrolls than pans
+        lo = max(self.canvas.data_yrange[0], lo)
+        hi = min(hi, self.canvas.data_yrange[1] + (self.canvas.data_yrange[1] - self.canvas.data_yrange[0]) * self._vmargin)
+
+        self.vrange.value = (lo, hi)
+        self.vrange_lock.value = True  # disable autoscaling when the user zooms in
+
+    def shift_horizontal(self, shift):
+        """
+        Move the horizontal range. The range always stays within safe values.
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/gui/comp/viewport.py:1618]: "range" is what defines the min and max value seen in the plot. The class uses this concept all around. See the .hrange and .vrange VAs. I'll keep the names as it is.
By "safe value" I meant a value which is within the complete data range. I've updated the comment.

File: src/odemis/gui/comp/viewport.py [hunk 10]
```diff
@@ -1554,6 +1549,121 @@ def on_canvas_motion(self, evt):
...
+        hi = lo + new_span
+
+        # Clamp the ranges - this functions differently for scrolls than pans
+        if lo < self.canvas.data_xrange[0]:
+            lo = self.canvas.data_xrange[0]
+        if hi > self.canvas.data_xrange[1]:
+            hi = self.canvas.data_xrange[1]
+
+        self.hrange.value = (lo, hi)
+        self.hrange_lock.value = True  # disable autoscaling when the user zooms in
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/gui/comp/viewport.py:1583]: Autoscaling of the range.

File: src/odemis/gui/comp/viewport.py [hunk 11]
```diff
@@ -1554,6 +1549,121 @@ def on_canvas_motion(self, evt):
...
+        # proportion of zoom_centre
+        prop = (centre - rng[0]) / span
+
+        # zoom around the centre point
+        new_span = span * (1 + scale)
+
+        lo = centre - prop * new_span
+        hi = lo + new_span
+
+        # Clamp the ranges - this functions differently for scrolls than pans
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/gui/comp/viewport.py:1576]: I now simplified and just explain separately in each function how we define the min/max values.

File: src/odemis/gui/comp/viewport.py [hunk 12]
```diff
@@ -1554,6 +1549,121 @@ def on_canvas_motion(self, evt):
...
+          becomes too large and gets clipped, the center might not be respected.
+        """
+        if not self.canvas.has_data():
+            return
+
+        rng = self.hrange.value
+        span = rng[1] - rng[0]
+
+        # proportion of zoom_centre
+        prop = (centre - rng[0]) / span
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/gui/comp/viewport.py:1568]: Is it clearer to say "relative position of the centre"?

File: src/odemis/gui/comp/viewport.py [hunk 13]
```diff
@@ -1554,6 +1549,121 @@ def on_canvas_motion(self, evt):
...
+        lo = max(self.canvas.data_yrange[0], lo)
+        hi = min(hi, self.canvas.data_yrange[1] + (self.canvas.data_yrange[1] - self.canvas.data_yrange[0]) * self._vmargin)
+
+        self.vrange.value = (lo, hi)
+        self.vrange_lock.value = True  # disable autoscaling when the user zooms in
+
+    def shift_horizontal(self, shift):
+        """
+        Move the horizontal range. The range always stays within safe values.
+        shift (float): amount to shift, in range units
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/gui/comp/viewport.py]: That's the unit of the range, whichever that unit is. Updated the doc to be slightly clearer. If you can suggest a better way to describe that, I would welcome it!

File: src/odemis/gui/comp/viewport.py [hunk 14]
```diff
@@ -1554,6 +1549,121 @@ def on_canvas_motion(self, evt):
...
+        """
+        # Move the horizontal axis
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/gui/comp/viewport.py:1644]: That's a left over comment. It's now replaced by the function name...

File: src/odemis/gui/comp/viewport.py [hunk 15]
```diff
@@ -1554,6 +1549,121 @@ def on_canvas_motion(self, evt):
...
+        # Make sure we don't go out of range
+        if lo < self.canvas.data_xrange[0]:
+            lo = self.canvas.data_xrange[0]
+            hi = self.canvas.data_xrange[0] + (self.canvas.display_xrange[1] - self.canvas.display_xrange[0])
+        if hi > self.canvas.data_xrange[1]:
+            hi = self.canvas.data_xrange[1]
+            lo = self.canvas.data_xrange[1] - (self.canvas.display_xrange[1] - self.canvas.display_xrange[0])
+
+        self.hrange.value = (lo, hi)
+        self.hrange_lock.value = True  # disable autoscaling when the user drags the scale
```
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/gui/comp/viewport.py:1663]: Here the scale meant the "legend". That's confusing. I've changed to "the user pans around".

File: src/odemis/gui/comp/viewport.py [hunk 16]
```diff
@@ -1422,38 +1430,46 @@ def on_char(self, evt):
...
         """ Scroll event for the bottom legend.
 
         Zooms the x-scale of the data and refreshes display
+        evt (MouseEvent)
         """
 
         rot = evt.GetWheelRotation() / evt.GetWheelDelta()
-        scale = -0.1 * rot
+        scale = -0.1 * rot  # ±10% of the range per scroll wheel step
         if evt.ShiftDown():  # softer
```
Comment (Sabrina85) [inline src/odemis/gui/comp/viewport.py:1445]: Can you add here similar to the above comments, that this means "smaller zoom"?

File: src/odemis/gui/comp/viewport.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Sabrina85) [reply src/odemis/gui/comp/viewport.py:1449]: Good that we have a comment now to clarify then :)

File: src/odemis/gui/comp/viewport.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Sabrina85) [reply src/odemis/gui/comp/viewport.py:1588]: Yes that sounds better. Can you add `relative position of the center in respect to xxxx` ?

File: src/odemis/gui/comp/viewport.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Sabrina85) [reply src/odemis/gui/comp/viewport.py:1600]: Is that similar to the auto button we have for auto BC?

**Review by Sabrina85 — CHANGES_REQUESTED** [review]
It was really worth to put the effort in the documentation. It is now much easier to follow :) I reviewed the last 3 commits. Let me know if I need to also look at the first 3 as there was a force push.

**Review by Sabrina85 — APPROVED** [review]
I approve the pr now as I am on holiday and it looked already good. I had only 2 small comments anyways.

File: src/odemis/gui/comp/viewport.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (pieleric) [reply src/odemis/gui/comp/viewport.py:1600]: Sort-of. It's a "lock" button. You can test it by loading a spectrum test data in the Odemis viewer.

**Pull request merged**